File: wrappers/uniffi/swift/Tests/AsdkTests/Http.swift
import Testing
import Foundation
import Swifter
@testable import Asdk

@Suite(.serialized) class HttpTests {

	let server: HttpServer

	init() async throws {
		self.server = HttpServer()
		try server.start(9003)
	}

	deinit {
		server.stop()
	}

	@Test func asyncCall() async throws {
		let client = MockHttpClient()

		let request = Asdk.HttpRequest(
			url: "http://localhost:9003/get",
			method: Asdk.HttpMethod.get,
			headers: ["accept": "application/json"],
			body: "{\"message\": \"are you ok?\"}"
		)

		let response = try await client.asyncCall(request: request)
		#expect(response.body == "{\"message\": \"ok\"}")
	}

	@Test func reqwestInsecureAsyncCall() async throws {
		self.server["/get"] = { (request: Swifter.HttpRequest) -> Swifter.HttpResponse in
			return .ok(
				.data(
                    "{\"message\": \"ok\"}".data(using: .utf8)!,
					contentType: "application/json"))
		}

		let client = try ReqwestHttpClient.insecure()
		let request = Asdk.HttpRequest(
			url: "http://localhost:9003/get",
			method: Asdk.HttpMethod.get,
			headers: ["accept": "application/json"],
			body: nil
		)
		let response = try await client.asyncCall(request: request)

		#expect("{\"message\": \"ok\"}" == response.body)
	}

	@Test func reqwestInsecureAsyncCallShouldThrowErrorOnSecureUrl() async throws {
		let client = try ReqwestHttpClient()
		let request = Asdk.HttpRequest(
			url: "http://localhost:9003/get",
			method: Asdk.HttpMethod.get,
			headers: ["accept": "application/json"],
			body: nil
		)

		do {
			_ = try await client.asyncCall(request: request)
			#expect(Bool(false), "Expected HttpAsyncCall error to be thrown")
		} catch let error as Asdk.Error {
			switch error {
			case .HttpAsyncCall(let message):
				#expect(message.contains("builder error for url"))
			default:
				#expect(Bool(false), "Unexpected Asdk.Error case: \(error)")
			}
		} catch {
			#expect(Bool(false), "Unexpected error type: \(error)")
		}
	}

}

final class MockHttpClient: Asdk.HttpClient {
	func asyncCall(request: Asdk.HttpRequest) async throws -> Asdk.HttpResponse {
		return Asdk.HttpResponse(
			statusCode: 200,
			headers: ["content-type": "application/json"],
			body: "{\"message\": \"ok\"}"
		)
	}
}

File: wrappers/uniffi/kotlin/src/test/kotlin/com/equs/http.kt
package com.equs

import com.equs.credentials.*
import kotlinx.coroutines.test.runTest
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.BeforeAll
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class Http {
    companion object {
        lateinit var mockServer: MockWebServer

        @JvmStatic
        @BeforeAll
        fun setup() {
            setJniLibPath()

            mockServer = MockWebServer()
            mockServer.start(9000)
        }

        @JvmStatic
        @AfterAll
        fun after_all() {
            mockServer.shutdown()
        }
    }

    @Test
    fun asyncCall() = runTest {
        val client = MockHttpClient()
        val request = HttpRequest(
            url = "http://localhost:1512",
            method = HttpMethod.GET,
            headers = mapOf(
                "Content-Type" to "application/json",
            ),
            body = "{\"message\": \"are you ok?\"}"
        )
        val response = client.asyncCall(request)

        assertEquals("{\"message\": \"ok\"}", response.body)
    }

    @Test
    fun reqwestInsecureAsyncCall() = runTest {
        mockServer.enqueue(
            MockResponse().setResponseCode(200).setBody("{\"message\": \"ok\"}")
                .setHeader("content-type", "application/json")
        )
        val client = ReqwestHttpClient.insecure()
        val request = HttpRequest(
            url = "http://localhost:9000",
            method = HttpMethod.GET,
            headers = mapOf(
                "accept" to "application/json",
            ),
            body = null
        )
        val response = client.asyncCall(request)

        assertEquals("{\"message\": \"ok\"}", response.body)
    }

    @Test
    fun reqwestInsecureAsyncCallShouldThrowErrorOnSecureUrl() = runTest {
        val client = ReqwestHttpClient()
        val request = HttpRequest(
            url = "http://localhost:9000/get",
            method = HttpMethod.GET,
            headers = mapOf(
                "accept" to "application/json",
            ),
            body = null
        )

        val exception = assertFailsWith<Exception.HttpAsyncCall> {
            client.asyncCall(request)
        }
        assertEquals(
            "v1=HTTP error: builder error for url (http://localhost:9000/get)",
            exception.message
        )
    }

    fun reqwestSecureAsyncCall() {
        TODO()
    }
}


class MockHttpClient : HttpClient {

    override suspend fun asyncCall(request: HttpRequest): HttpResponse {
        assertEquals(request.body, "{\"message\": \"are you ok?\"}")
        assertEquals(request.method, HttpMethod.GET)
        return HttpResponse(
            statusCode = 200.toUShort(),
            headers = mapOf(
                "Content-Type" to "application/json",
            ),
            body = "{\"message\": \"ok\"}"
        )
    }

}